test/wallet-test.js

```diff
@@ -825,6 +837,128 @@ describe('Wallet', function() {
+
+    // m'/44'/0' Oops, wrong one.
+    const bitcoinKey = bip44Key.derive(0, true);
+    const xpub = bitcoinKey.xpubkey(network.type);
+    const key = HDPublicKey.fromBase58(xpub, network);
+
+    const wallet = await wdb.create({
+      watchOnly: true
+    });
+
+    let err;
+    try {
+      await wallet.createAccount({
+        name: 'foo',
+        accountKey: key
+      });
+    } catch (e) {
+      err = e;
+    }
+
+    assert(err);
+    assert.strictEqual(err.message, 'Expected key with account depth.');
+  });
+
+  it('should fail w/ watch-only with incorrect account hd pubkey', async() => {
+    const network = Network.get('regtest');
+    const mnemonic = Mnemonic.fromPhrase(PHRASE1);
+
+    // m'
+    const priv = HDPrivateKey.fromMnemonic(mnemonic);
+
+    // m'/44'
+    const bip44Key = priv.derive(44, true);
+
+    // m'/44'/0'
+    const bitcoinKey = bip44Key.derive(0, true);
+
+    // m'/44'/0'/0'
+    const accountKey = bitcoinKey.derive(0, true);
+    const xpub = accountKey.xpubkey(network.type);
+    const key = HDPublicKey.fromBase58(xpub, network);
+
+    // m'/44'/0'/5' Oops, wrong one.
+    const account5Key = bitcoinKey.derive(5, true);
+    const xpub5 = account5Key.xpubkey(network.type);
+    const key5 = HDPublicKey.fromBase58(xpub5, network);
+
+    const wallet = await wdb.create({
+      watchOnly: true,
+      accountKey: key
+    });
+
+    let err;
+    try {
+      await wallet.createAccount({
+        name: 'foo',
+        accountKey: key5
+      });
+    } catch (e) {
+      err = e;
+    }
+
+    assert(err);
+    assert.strictEqual(err.message, 'Expected account key 1.');
+  });
+
+  it.skip('should fail w/ watch-only with master key mismatch', async() => {
+    const network = Network.get('regtest');
+    const mnemonic1 = Mnemonic.fromPhrase(PHRASE1);
+    const mnemonic2 = Mnemonic.fromPhrase(PHRASE2);
+
+    // m'
+    const priv1 = HDPrivateKey.fromMnemonic(mnemonic1);
+    const priv2 = HDPrivateKey.fromMnemonic(mnemonic2);
+
+    // m'/44'
+    const bip44Key1 = priv1.derive(44, true);
+    const bip44Key2 = priv2.derive(44, true);
+
+    // m'/44'/0'
+    const bitcoinKey1 = bip44Key1.derive(0, true);
+    const bitcoinKey2 = bip44Key2.derive(0, true);
+
+    // m'/44'/0'/0'
+    const accountKey1 = bitcoinKey1.derive(0, true);
+    const xpub1 = accountKey1.xpubkey(network.type);
+    const key1 = HDPublicKey.fromBase58(xpub1, network);
+
+    // m'/44'/0'/0'
+    const accountKey2 = bitcoinKey2.derive(0, true);
+    const xpub2 = accountKey2.xpubkey(network.type);
+    const key2 = HDPublicKey.fromBase58(xpub2, network);
+
+    const wallet = await wdb.create({
+      watchOnly: true,
+      accountKey: key1
+    });
+
+    let err;
+    try {
+      // Oops, wrong one.
+      await wallet.createAccount({
+        name: 'foo',
+        accountKey: key2
+      });
+    } catch (e) {
+      err = e;
+    }
+
+    assert(err);
+    assert.strictEqual(err.message, 'Master key mismatch.');
```

Comment:
If we're testing collisions maybe the error should say "key already exists at index..."

Reply:
It's testing if the key is derived from the same parent. 

The collision of the `accountDepth` is one of  the potential symptoms of mismatching.

Reply:
The other issue is that if account 0, 1, 2 are from key-a, and account 3 is from key-b, it's a mismatch issue again, and an attempt to sign account 3 with key-a would fail.

Reply:
The `parentFingerPrint` can be used to validate that the accounts all derived from the same key to avoid those issues.